File: golang/2487_remove_nodes/remove_nodes.go
/*
 * @Date: 2024-01-03
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-01-03
 * @FilePath: /algorithm/golang/2487_remove_nodes/remove_nodes.go
 */

// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverse(head *ListNode) *ListNode {
	dummy := &ListNode{}
	for head != nil {
		p := head
		head = head.Next
		p.Next = dummy.Next
		dummy.Next = p
	}
	return dummy.Next
}

func removeNodes(head *ListNode) *ListNode {
	head = reverse(head)
	for p := head; p.Next != nil; {
		if p.Val > p.Next.Val {
			p.Next = p.Next.Next
		} else {
			p = p.Next
		}
	}
	return reverse(head)
}

func isSameList(l1, l2 *ListNode) bool {
	for l1 != nil && l2 != nil {
		if l1.Val != l2.Val {
			return false
		}
		l1 = l1.Next
		l2 = l2.Next
	}
	if l1 != nil || l2 != nil {
		return false
	}
	return true
}

func main() {
	tests := []struct {
		head *ListNode
		ans  *ListNode
	}{
		{
			&ListNode{5, &ListNode{2, &ListNode{13, &ListNode{3, &ListNode{8, nil}}}}},
			&ListNode{13, &ListNode{8, nil}},
		},
		{
			&ListNode{1, &ListNode{1, &ListNode{1, &ListNode{1, nil}}}},
			&ListNode{1, &ListNode{1, &ListNode{1, &ListNode{1, nil}}}},
		},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, true, isSameList(removeNodes(test.head), test.ans), index)
	}
}

File: cpp/2487_remove_nodes/remove_nodes.cpp
/*
 * @Date: 2024-01-03
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-01-03
 * @FilePath: /algorithm/cpp/2487_remove_nodes/remove_nodes.cpp
 */

// Definition for singly-linked list.
#include <cassert>
#include <tuple>
#include <vector>

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode *reverse(ListNode *head) {
    ListNode dummy;
    while (head != nullptr) {
      ListNode *p = head;
      head = head->next;
      p->next = dummy.next;
      dummy.next = p;
    }
    return dummy.next;
  }

  ListNode *removeNodes(ListNode *head) {
    head = reverse(head);
    for (ListNode *p = head; p->next != nullptr;) {
      if (p->val > p->next->val) {
        p->next = p->next->next;
      } else {
        p = p->next;
      }
    }
    return reverse(head);
  }
};

bool is_same_list(ListNode *l1, ListNode *l2) {
  while (l1 != NULL && l2 != NULL) {
    if (l1->val != l2->val) {
      return false;
    }
    l1 = l1->next;
    l2 = l2->next;
  }

  // 如果其中一个链表遍历完了，但另一个链表还没有遍历完，那么它们不相等。
  if (l1 != NULL || l2 != NULL) {
    return false;
  }

  return true;
}

int main() {
  std::vector<std::tuple<ListNode *, ListNode *>> tests{
      {
          new ListNode{5, new ListNode{2, new ListNode{13, new ListNode{3, new ListNode{8}}}}},
          new ListNode{13, new ListNode{8}},
      },
      {
          new ListNode{1, new ListNode{1, new ListNode{1, new ListNode{1}}}},
          new ListNode{1, new ListNode{1, new ListNode{1, new ListNode{1}}}},
      },
  };

  for (auto &[l1, ans] : tests) {
    assert(is_same_list(Solution().removeNodes(l1), ans));
  }
}